### PR TITLE
feat(overlay): add support for flexible connected positioning

### LINKED ISSUE #2725: md-menu is defaulted to `max-height: calc(100vh + 48px);` and goes off page
#### Bug, feature request, or proposal:
Bug

#### What is the expected behavior?
I expect that the menu would size to stay on the screen and scroll when it has a lot of items in it.

#### What is the current behavior?
Currently, it will just go off the bottom of the page and I won't be able to see a lot of items.

image
This is the bottom of my browser and there are more items below it.

#### What are the steps to reproduce?
Add a menu with a lot of buttons. Don't put the menu at the top of the page. Open the menu, it will go down below the bottom of the page.

http://plnkr.co/edit/FPHeqOjRHpJOBNGn5dvz?p=preview

#### What is the use-case or motivation for changing an existing behavior?
If I want a bunch of items in a list.

#### Which versions of Angular, Material, OS, browsers are affected?
Angular: 2.3.1
Material: 2.0.0-alpha.11-3
Mac
Chrome

And I assume it's on most others.

#### Is there anything else we should know?
It looks like this is happening in https://github.com/angular/material2/blob/master/src/lib/menu/menu.scss#L16. So having it be a full viewport means it's guaranteed to go below the page.

### LINKED ISSUE #5267: Overlay connected position strategy should have an option to push the contents into the screen.
#### Bug, feature request, or proposal:
Overlay-connected-position strategy should have an option to push the contents into the screen.

Overlay-connected-position strategy allows setting something like "put the overlay element horizontally centered", so you could have something like this:

```
 ________________
|                |
|           o    |
|           ^    |
|          ███   |
|________________|
```

This is nice, but there is one problem, what if the element is in a position where no matter where you put the overlay contents, they will always be rendered off the screen?

```
 ________________
|                |
|           o    |
|           ^    |
|    ████████████|██
|________________|
```

#### What is the expected behavior?
Given an option in the connected-position-strategy, push the element into the screen by the largest amount of pixels that it goes off-screen. For example:

* If it goes 20px off screen to the right but there are 50px available to the left before going off screen again, then push it by 20px.
* If it goes 10px off screen to the right but there are only 5px available to the left before going off screen again, then push it by 5px.
* If it goes 10px off screen to the right but there are is already 20px offscreen to the left, then do nothing.

```
 ________________
|                |
|           o    |
|           ^    |
| ███████████████|
|________________|
```

#### What is the current behavior?
The position strategy gets the amount of pixels that it goes off-screen, and subtracts them from the overlay contents to get the "visible pixels", then multiplies HxW to get the area and selects the position with the largest visible area from the list of fallback positions even if it renders off screen.

#### What is the use-case or motivation for changing an existing behavior?
Tooltip-like components that render on hover or focus will render after the element. Normally this is not an issue on desktops or tablets because there is a lot of space, but on tiny phones like an iPhone4, this doesn't look good.

Note that we can't use tooltips because we want buttons, links, and other stuff.

### LINKED ISSUE #6534: Support flexible overlay size with connected position strategy
This issue tracks the high-level feature of supporting overlays that are automatically an appropriate size based on their position and viewport bounds. This is accomplished by creating a bounding container element that constrains the size of the overlay panel. 

In-progress work is: https://github.com/jelbourn/material2/blob/better-connected-position/src/lib/core/overlay/position/better-connected-position-strategy.ts

### PR DESCRIPTION
**Pull request by crisbeto**
* Adds the `FlexibleConnectedPositionStrategy` that builds on top of the `ConnectedPositionStrategy` adds the following:
  * The ability to have overlays with flexible sizing.
  * Being able to push overlays into the viewport if they don't fit.
  * Having a margin between the overlay and the viewport edge.
  * Being able to assign weights to the overlay positions.
* Refactors the `ConnectedPositionStrategy` to use the `FlexibleConnectedPositionStrategy` in order to avoid breaking API changes.
* Switches all of the components to the new position strategy.
* Adds an API to the `OverlayRef` that allows for the consumer to wrap the pane in a div. This is a common requirement between the `GlobalPositionStrategy` and the `FlexibleConnectedPositionStrategy`, and it's easier to keep track of the elements when the attaching and detaching is done in the `OverlayRef`.

Fixes #6534.
Fixes #2725.
Fixes #5267.

### CODE REVIEW

**Comment by isaacplmann** [comment]
I'm excited for this to land! Thanks for the work @crisbeto.

File: src/cdk/overlay/position/flexible-connected-position-strategy.ts [hunk 1]
```diff
@@ -0,0 +1,917 @@
...
+  private _lastPosition: ConnectedPosition;
+
+  /** Subject that emits whenever the position changes. */
+  private _positionChange = new Subject<ConnectedOverlayPositionChange>();
+
+  /** Subscription to viewport size changes. */
+  private _resizeSubscription = Subscription.EMPTY;
+
+  /** Observable sequence of position changes. */
+  positionChange: Observable<ConnectedOverlayPositionChange> = this._positionChange.asObservable();
```
Comment (jelbourn) [inline src/cdk/overlay/position/flexible-connected-position-strategy.ts]: `positionChanges`?

File: src/cdk/overlay/position/flexible-connected-position-strategy.spec.ts [hunk 2]
```diff
@@ -0,0 +1,1333 @@
...
+          overlayX: 'start',
+          originY: 'bottom',
+          originX: 'start'
+        }]);
+
+      attachOverlay({positionStrategy});
+
+      const overlayRect = overlayRef.overlayElement.getBoundingClientRect();
+      expect(Math.floor(overlayRect.height)).toBe(OVERLAY_HEIGHT - 10);
+      expect(Math.floor(overlayRect.bottom)).toBe(viewport.getViewportSize().height);
```
[the remark this answers is not in the document]
Comment (crisbeto) [reply src/cdk/overlay/position/flexible-connected-position-strategy.spec.ts]: I don't think so, we're positioning relative to the viewport, not the page.

File: src/cdk/overlay/position/flexible-connected-position-strategy.ts [hunk 3]
```diff
@@ -0,0 +1,915 @@
...
+
+    // Use a transform to apply the offsets. We do this because the `center` positions rely on
+    // being in the normal flex flow and setting a `top` / `left` at all will completely throw
+    // off the position. We also can't use margins, because they won't have an effect in some
+    // cases where the element doesn't have anything to "push off of". Finally, this works
+    // better both with flexible and non-flexible positioning.
+    let transformString = '';
+
+    if (position.offsetX) {
+      transformString += `translateX(${position.offsetX}px)`;
```
Comment (mmalerba) [inline src/cdk/overlay/position/flexible-connected-position-strategy.ts]: prefix with `' '` to prevent potential future bugs (the `trim` will just remove it anyways). or use an array and `.join(' ')`

File: src/cdk/overlay/position/flexible-connected-position-strategy.ts [hunk 4]
```diff
@@ -0,0 +1,915 @@
...
+      throw Error('This position strategy is already attached to an overlay');
+    }
+
+    overlayRef.hostElement.classList.add('cdk-overlay-connected-position-bounding-box');
+
+    this._overlayRef = overlayRef;
+    this._boundingBox = overlayRef.hostElement!;
+    this._pane = overlayRef.overlayElement;
+    this._resizeSubscription.unsubscribe();
+    this._resizeSubscription = this._viewportRuler.change().subscribe(() => this.apply());
```
[the remark this answers is not in the document]
Comment (crisbeto) [reply src/cdk/overlay/position/flexible-connected-position-strategy.ts]: We need to do it inside the `disposed` and I feel like introducing another `Subject` for this one case would be overkill.

File: src/cdk/overlay/position/flexible-connected-position-strategy.ts [hunk 5]
```diff
@@ -0,0 +1,915 @@
...
+      // position.
+      if (!fallback || fallback.overlayFit.visibleArea < overlayFit.visibleArea) {
+        fallback = {overlayFit, overlayPoint, originPoint, position: pos, overlayRect};
+      }
+    }
+
+    // If there are any positions where the overlay would fit with flexible dimensions, choose the
+    // one that has the greatest area available modified by the position's weight
+    if (flexibleFits.length) {
+      let bestFit: FlexibleFit | null = null;
```
[the remark this answers is not in the document]
Comment (crisbeto) [reply src/cdk/overlay/position/flexible-connected-position-strategy.ts]: I'm not sure that it's the same functionality. If we did that, the first fit would have a score of -1 (unless we also moved out the score calculation) which would make it impossible to select.

File: src/cdk/overlay/position/flexible-connected-position-strategy.ts [hunk 6]
```diff
@@ -0,0 +1,915 @@
...
+      this._applyPosition(bestFit!.position, bestFit!.origin);
+      return;
+    }
+
+    // When none of the preferred positions fit within the viewport, take the position
+    // that went off-screen the least and attempt to push it on-screen.
+    if (this._canPush) {
+      // TODO(jelbourn): after pushing, the opening "direction" of the overlay might not make sense.
+      this._isPushed = true;
+      this._applyPosition(fallback!.position, fallback!.originPoint);
```
[the remark this answers is not in the document]
Comment (crisbeto) [reply src/cdk/overlay/position/flexible-connected-position-strategy.ts]: I don't think so. It should've been assigned here, unless we exited already https://github.com/angular/material2/pull/9153/files#diff-be0f6cfbf065a6175e266e5bc8cd5213R219.

**Comment by crisbeto** [comment]
Addressed the feedback @mmalerba @jelbourn.

**Comment by jelbourn** [comment]
I ran an early presubmit on this and it resulted in a fair number of failures. I'll try to look into them today to figure out if it's just their tests or something in the PR.

(I suspect it's their tests depending on a specific DOM structure)

**Comment by crisbeto** [comment]
FWIW, I was waiting for the final review to look into a couple of unit test failures on Safari, but IIRC they were only for the centered position + offset case which isn't super common.

File: src/cdk/overlay/position/flexible-connected-position-strategy.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (mmalerba) [reply src/cdk/overlay/position/flexible-connected-position-strategy.ts]: you'd obviously need to set the initial score correctly as well

File: src/cdk/overlay/position/flexible-connected-position-strategy.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (mmalerba) [reply src/cdk/overlay/position/flexible-connected-position-strategy.ts]: It looks like its only set inside that for loop, so if `this._perferredPositions == []` it would not be set

File: src/cdk/overlay/position/flexible-connected-position-strategy.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (crisbeto) [reply src/cdk/overlay/position/flexible-connected-position-strategy.ts]: That means moving that long line outside the loop though.

**Comment by jelbourn** [comment]
Presubmit failure number 1: one app is somehow getting into a state where a call to `apply` is happening after the overlay is disposed (via some kind of async action). This manifests as `_boundingBox` being null when trying to read `style` from it. 

Adding an `_isDisposed` property and aborting `apply` based on that seems to fix it

**Comment by crisbeto** [comment]
Sorted out the `apply` after `dispose` issue.

**Comment by mmalerba** [comment]
@crisbeto needs rebase

**Comment by jelbourn** [comment]
Summary of remaining failures:
* Screenshot test failures where an autocomplete overlay is a few hundred pixels to the right of where it should be (multiple projects). We should double check the position config for autocomplete
* Screenshot test failures where menus are about 10px to the left of where they should be (multiple projects)
* Screenshot failures where the entire app UI is missing (not totally sure if this is related, needs more investigation)
* Unit test failure where the innerText of the overlay container used to be something but now is empty (possibly a timing issue)
* Unit test failure where `overlayContainer.querySelector('.mat-menu-panel')` is returning null when formerly it didn't (again might be due to timing)

I'll try to debug these tomorrow, but you might have ideas in the meantime.

**Comment by crisbeto** [comment]
The first two sound like something is definitely wrong on our end, but I'm not sure what it would be off the top of my head. As far as I remember the logic for a non-flexible overlay is more or less the same. I've also ported all the unit tests from the old positioning into the new one and they passed with zero changes.

**Comment by jelbourn** [comment]
Turns out that first one was actually a select. It was caused by the old version of select we have in Google (called mat-deprecated-select) that some teams are still using. This was leftover from when we switched select to be inside form-field. I was able to update the old code to resolve the issue (the `_onAttached` function just needed to be updated to match the current implementation)

**Comment by jelbourn** [comment]
So, for the menu being slightly off, I think I found what's wrong.

The case is where the menu-trigger is close to the right side of the screen.

The overlay-container is set to `width: 100%` rather than `100vw`, with the difference being the width of the scrollbar. When we set `right` (on the bounding box), it's relative to the viewport without the scrollbar (`width: 100%`), but `viewportRuler` uses `innerWidth`, which includes the scrollbar.

We _probably_ don't want to change overlay-container since it would likely affect downstream apps. We should be able to manually account for the scrollbar in the calculations. We could alternatively use `body.getBoundingClientRect().right` instead of the value we're currently getting from `ViewportRuler` (updating `ViewportRuler` would probably also be risky).

**Comment by crisbeto** [comment]
The `100%` vs `100vw` thing came up when we first introduced the `bottom`/`right` behavior, and if I remember correctly we tried making the container `100vw` which ended up breaking a lot of other things. The workaround was to use the `documentElement.clientWidth` instead. We still have it when assigning the overlay pane styling, but we forgot to account for it when calculating the bounding box styles. I've rebased it and pushed a fix (it's only in the `_getNarrowedViewportRect` method if you don't want to re-review the whole thing).

**Comment by crisbeto** [comment]
Done @jelbourn.
